### PR TITLE
PTI-298: CourtConvictions - Add admin/public add/edit/details + Penalties add/edit/details Components

### PR DESCRIPTION
**Pull request by NickPhura**
There are some nice-to-have changes I want to make in separate PRs after this lands (these aren't specific to court convictions necessarily):
 - Update a bunch of the `spans` to `labels`
 - Update the scss for the details pages, so they pull from a common file.
   - As part of this I can also look at moving it to use more Bootstrap classes if possible, etc.

## Commit 1:
- Api changes
  - Add court convictions controllers
  - Update several record types to have Penalties
    - In records that get Penalties, it replaces the old Outcome Status/Description fields.
 - Update models accordingly

## Commit 2:
- Global changes
  - Added two new directives
    - One expands textarea's as you type, with min/max sizes
    - The other enforces only digits in a field, with options to allow 1 decimal or not

## Commit 3:
- Frontend changes (admin, public, common)
  - Add court convictions add/edit/details/public details components
  - Add Penalties add/edit/details/public details components
  - Update several record types to have Penalties
    - In the records that get Penalties, it replaces the old Outcome Status/Description fields.

The frontend changes aren't broken down as far as I'd normally like, but I originally wrote this having it broken down by features, which got too complicated when I was rebasing, so I just squashed all of the front-end code into 1 commit.

### CODE REVIEW

File: angular/projects/admin-nrpti/src/app/header/header.component.ts [hunk 1]
```diff
@@ -25,8 +25,8 @@ export class HeaderComponent {
 
   constructor(public factoryService: FactoryService, public router: Router) {
     router.events.subscribe(() => {
-      this.isAuthenticated = this.factoryService.isAuthenticated();
-      this.welcomeMsg = this.factoryService.getWelcomeMessage();
+      // this.isAuthenticated = this.factoryService.isAuthenticated();
+      // this.welcomeMsg = this.factoryService.getWelcomeMessage();
     });
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/header/header.component.ts:30]: What's going on here?

File: angular/projects/common/src/app/penalty/penalty-detail-admin/penalty-detail.component.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  //       this.data.map((penalty: Penalty) => {
+  //         return {
+  //           penaltyString: penalty.buildPenaltyValueString(),
+  //           type: penalty.type,
+  //           description: penalty.description
+  //         };
+  //       })) ||
+  //     [];
+  // }
+}
```
Comment (marklise) [inline angular/projects/common/src/app/penalty/penalty-detail-admin/penalty-detail.component.ts:53]: Remove

File: angular/projects/common/src/app/penalty/penalty-detail-public/penalty-detail.component.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  // prepTextFields() {
+  //   if (this.data && this.data.length) {
+  //     this.preparedPenaltyStrings = [];
+
+  //     this.data.forEach(penalty => {
+  //       this.preparedPenaltyStrings.push(CommonUtils.buildPenaltyString(new Penalty(penalty)));
+  //     });
+  //   }
+  // }
+
```
Comment (marklise) [inline angular/projects/common/src/app/penalty/penalty-detail-public/penalty-detail.component.ts:50]: Remove - and any other occurrences of commented out code.

File: angular/projects/public-nrpti/src/app/records/court-convictions/court-conviction-detail/court-conviction-detail.component.html [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        </li>
+        <!-- <li
+          class="nav-item"
+          [ngClass]="{ active: isTabActive('location') }"
+          (click)="activateTab('location')"
+          title="Court conviction location details"
+        >
+          <i class="nav-item-icon material-icons">room</i>
+          <a class="nav-item-value">Location</a>
+        </li> -->
```
Comment (marklise) [inline angular/projects/public-nrpti/src/app/records/court-convictions/court-conviction-detail/court-conviction-detail.component.html:22]: Remove.

**Comment by NickPhura** [comment]
> 
> 
> Lots of commented out code.
> 
> Visuals of the penalties look odd at various times. Prefer the use of bootstrap classes than custom css. I wonder about keeping it simple and leaving the input boxes and drop downs separate? Also the text box on the right doesn't align correctly.
> 
> image

Yeah I commented above about using more bootstrap classes, as well as some other cleanup.  But I don't think Ill have enough time pre-launch.

The input box doesn't line up because its a textarea, and apparently has its own styling.  I played around with altering the line-height for it, and it required some weird custom value to make it visually align with the inputs next to it.  I changed the initial height to be larger, so that it doesn't look like it was supposed to be matching, but was off.

**Comment by marklise** [comment]
> In regards to the textareas some more: It looks like chrome and firefox render them slightly differently as well (chromes have slightly less height than firefoxes), so Its probably not super easy to make them match exactly across browsers.

Maybe set it to a multiline input instead of a draggable resizer?

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.html [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        (linksChanged)="links = $event"
+        (documentsChanged)="documents = $event"
+      ></app-document-link-staging>
+    </section>
+
+    <section>
+      <div class="row mb-4">
+        <div class="col-2">
+          <h4 class="blue-header mb-0">NRCED Site Content</h4>
+          <br />
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.html:166]: These and all the other BR's are unnecessary.

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.html [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            name="publishNrced"
+            id="publishNrced"
+            formControlName="publishNrced"
+            (change)="togglePublish($event, 'nrced')"
+            [checked]="myForm.controls.publishNrced.value"
+            ngDefaultControl
+          >
+            {{ myForm.controls.publishNrced.value ? 'Published' : 'Unpublished' }}
+          </mat-slide-toggle>
+          <br />
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.html:180]: Boostrap classes can be used for padding instead.

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.scss [hunk 7]
```diff
@@ -0,0 +1,28 @@
+@import "assets/styles/base/base.scss";
+@import "assets/styles/components/add-edit.scss";
+
+.grey-subtext {
+  color: grey;
+  font-size: 15px;
+  font-style: italic;
+  margin-left: 20px;
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.scss:8]: Better to use a bootstrap class for margins.

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+  public courtConvictionSubtypes = Picklists.courtConvictionSubtypePicklist;
+
+  // Documents
+  public documents = [];
+  public links = [];
+  public documentsToDelete = [];
+
+  constructor(
+    public route: ActivatedRoute,
+    public router: Router,
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts:45]: Should these be private variables instead?

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+        description: penaltyFormGroup.description
+      });
+    });
+
+    return penalties;
+  }
+
+  navigateToDetails() {
+    this.router.navigate(['records', 'court-convictions', this.currentRecord._id, 'detail']);
+  }
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts:249]: I don't see this being used anywhere.

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+
+  async submit() {
+    // TODO
+    // _epicProjectId
+    // _sourceRefId
+    // _epicMilestoneId
+    // legislation
+    // projectName
+
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts:272]: These look like just copy/paste cruft.  Are they actually TODO's?

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+        this.recordUtils.parseResForErrors(res);
+        const docResponse = await this.recordUtils.handleDocumentChanges(
+          this.links,
+          this.documents,
+          this.documentsToDelete,
+          res[0][0].object._id,
+          this.factoryService
+        );
+
+        console.log(docResponse);
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts:371]: Remove

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+        this.recordUtils.parseResForErrors(res);
+        const docResponse = await this.recordUtils.handleDocumentChanges(
+          this.links,
+          this.documents,
+          this.documentsToDelete,
+          this.currentRecord._id,
+          this.factoryService
+        );
+
+        console.log(docResponse);
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-add-edit/court-conviction-add-edit.component.ts:405]: Remove

File: angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-detail/court-conviction-detail.component.html [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+            </button>
+          </div>
+        </div>
+        <div class="grid-section-sub-header">
+          <span class="grid-section-sub-header-text">Basic Information</span>
+        </div>
+        <div class="grid-section section-1">
+          <div class="grid-item__row">
+            <span class="grid-item-name">Record Name</span>
+            <span class="grid-item-value__col">{{ (data && data._master.recordName) || '-' }}</span>
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/court-convictions/court-conviction-detail/court-conviction-detail.component.html:25]: These and other areas, we should be referencing from those pages that are label based, not span.

File: angular/projects/admin-nrpti/src/app/records/records-list/records-list.component.html [hunk 14]
```diff
@@ -18,6 +18,7 @@
           <button ngbDropdownItem (click)="add('administrativeSanction')">Administrative Sanction</button>
           <button ngbDropdownItem (click)="add('agreement')">Agreement</button>
           <button ngbDropdownItem (click)="add('certificate')">Certificate</button>
+          <button ngbDropdownItem (click)="add('courtConviction')">Court Conviction</button>
           <button ngbDropdownItem (click)="add('constructionPlan')">Construction Plan</button>
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/records-list/records-list.component.html:22]: We should probably reverse these based on alphasort

File: angular/projects/admin-nrpti/src/app/records/records-list/records-list.component.ts [hunk 15]
```diff
@@ -133,6 +133,10 @@ export class RecordsListComponent implements OnInit, OnDestroy {
         {
           displayName: 'ManagementPlan',
           fieldName: '_schemaName'
+        },
+        {
+          displayName: 'CourtConviction',
+          fieldName: '_schemaName'
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/records-list/records-list.component.ts:139]: Reverse order.

File: angular/projects/admin-nrpti/src/app/records/records-list/records-list.component.ts [hunk 16]
```diff
@@ -326,6 +330,9 @@ export class RecordsListComponent implements OnInit, OnDestroy {
       case 'managementPlan':
         this.router.navigate(['records', 'management-plans', 'add']);
         break;
+      case 'courtConviction':
+        this.router.navigate(['records', 'court-convictions', 'add']);
+        break;
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/records-list/records-list.component.ts:335]: Reverse order.

File: angular/projects/admin-nrpti/src/app/records/records-resolver.ts [hunk 17]
```diff
@@ -25,7 +25,8 @@ export class RecordsResolver implements Resolve<Observable<object>> {
       'AdministrativeSanction',
       'Warning',
       'ConstructionPlan',
-      'ManagementPlan'
+      'ManagementPlan',
+      'CourtConviction'
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/records-resolver.ts:29]: Reverse order.

File: angular/projects/admin-nrpti/src/app/records/records-routing.module.ts [hunk 18]
```diff
@@ -701,6 +751,7 @@ const routes: Routes = [
     WarningResolver,
     ConstructionPlanResolver,
     ManagementPlanResolver,
+    CourtConvictionResolver,
```
Comment (marklise) [inline angular/projects/admin-nrpti/src/app/records/records-routing.module.ts:754]: Reverse order.

File: angular/projects/common/src/app/entity/entity-detail/entity-detail.component.html [hunk 19]
```diff
@@ -1,6 +1,6 @@
 <div class="grid-section section-1">
   <div class="grid-item">
-    <label class="grid-item-name">Entity Type</label>
+    <span class="grid-item-name">Entity Type</span>
```
Comment (marklise) [inline angular/projects/common/src/app/entity/entity-detail/entity-detail.component.html:3]: Shouldn't these be made into spans?  Is it the grid layout that is causing the problem?

**Pull request merged**